File: nes Watch App/Emulator/NES.swift
import CoreGraphics

final class NES {
    private let bus = Bus()
    private let cpu = CPU6502()
    private let ppu = PPU()

    private var cartridge: Cartridge?

    init() {
        cpu.bus = bus
        bus.cpu = cpu
        bus.ppu = ppu
    }

    func loadCartridge(_ cartridge: Cartridge) {
        self.cartridge = cartridge
        bus.cartridge = cartridge
        ppu.connectCartridge(cartridge)
        reset()
    }

    func reset() {
        cpu.reset()
    }

    func stepFrame() {
        guard cartridge != nil else { return }
        ppu.resetFrameState()
        while !ppu.frameComplete {
            let cycles = cpu.step()
            if cycles > 0 {
                for _ in 0..<(cycles * 3) {
                    ppu.tick()
                    if ppu.nmiRequested {
                        cpu.nmi()
                    }
                }
            }
        }
    }

    func currentFrameImage() -> CGImage? {
        return ppu.frameBuffer.makeImage()
    }

    func setButton(_ button: Controller.Button, pressed: Bool) {
        bus.controller.setButton(button, pressed: pressed)
    }
}

File: nes Watch App/Core/src/nes.c
#include "../include/nesc.h"

#include <stdlib.h>
#include <string.h>

#include "../include/nes_internal.h"

NESRef nes_create(void) {
    NES *nes = (NES *)calloc(1, sizeof(NES));
    if (!nes) {
        return NULL;
    }
    apu_init(&nes->apu);
    nes->bus.cpu = &nes->cpu;
    nes->bus.ppu = &nes->ppu;
    nes->bus.apu = &nes->apu;
    cpu_init(&nes->cpu);
    nes->cpu.bus = &nes->bus;
    return nes;
}

void nes_destroy(NESRef nes) {
    if (!nes) {
        return;
    }
    cartridge_free(&nes->cart);
    free(nes);
}

bool nes_load_rom(NESRef nes, const uint8_t *data, size_t size) {
    if (!nes) {
        return false;
    }
    cartridge_free(&nes->cart);
    memset(&nes->cart, 0, sizeof(nes->cart));
    if (!cartridge_load(&nes->cart, data, size)) {
        cartridge_free(&nes->cart);
        return false;
    }
    nes->bus.cartridge = &nes->cart;
    ppu_connect_cartridge(&nes->ppu, &nes->cart);
    nes->hasCart = true;
    nes_reset(nes);
    return true;
}

void nes_reset(NESRef nes) {
    if (!nes) {
        return;
    }
    cpu_reset(&nes->cpu);
}

void nes_step_frame(NESRef nes) {
    if (!nes || !nes->hasCart) {
        return;
    }
    ppu_reset_frame(&nes->ppu);
    while (!nes->ppu.frameComplete) {
        int cycles = cpu_step(&nes->cpu);
        if (cycles > 0) {
            for (int i = 0; i < cycles * 3; i++) {
                ppu_tick(&nes->ppu);
                if (nes->ppu.nmiRequested) {
                    cpu_nmi(&nes->cpu);
                }
            }
        }
    }
}

const uint32_t *nes_framebuffer(NESRef nes) {
    if (!nes) {
        return NULL;
    }
    return nes->ppu.frameBuffer.pixels;
}

int nes_framebuffer_width(void) { return NES_WIDTH; }
int nes_framebuffer_height(void) { return NES_HEIGHT; }

void nes_set_button(NESRef nes, uint8_t button, bool pressed) {
    if (!nes) {
        return;
    }
    controller_set_button(&nes->bus.controller, button, pressed);
}

float nes_apu_next_sample(NESRef nes, double sample_rate) {
    if (!nes) {
        return 0.0f;
    }
    return apu_next_sample(&nes->apu, sample_rate);
}
